8301136: Improve unlink() and unlink_all() of ResourceHashtableBase

navyxliu opened the pull request:
1. Apply the same idea of JDK-8300184 to unlink().
2. Because ResourceHashtableBase doesn't support copy assignment, client of it has to purge all elements first when it needs to assign it. We would like provide a specialized version called 'unlink_all()'.  We don't need to update each node's _next in this case. We only nullify all buckets. 
3. This patch also provides a specialized version of unlink_all() for destructor. We don't even update buckets. it's dead anyway.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301136](https://bugs.openjdk.org/browse/JDK-8301136): Improve unlink() and unlink_all() of ResourceHashtableBase


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12213/head:pull/12213` \
`$ git checkout pull/12213`

Update a local copy of the PR: \
`$ git checkout pull/12213` \
`$ git pull https://git.openjdk.org/jdk pull/12213/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12213`

View PR using the GUI difftool: \
`$ git pr show -t 12213`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12213.diff">https://git.openjdk.org/jdk/pull/12213.diff</a>

</details>

iklam commented on src/hotspot/share/utilities/resourceHash.hpp:
```diff
@@ -258,25 +248,59 @@ class ResourceHashtableBase : public STORAGE {
     }
   }
 
+  // unlink_all() is a specialized version of unlink() when we decide to remove all elements.
+  // It can not replace unlink(ITER* iter) if user-provided iter releases key/value
+  void unlink_all() {
+    Node** bucket = table();
+    const unsigned sz = table_size();
+
+    while (_number_of_entries > 0 && bucket < bucket_at(sz)) {
+      Node* node = *bucket;
```
Instead of having two loops for unlinking, I would suggest using the same pattern as iterate/iterate_all so you can share the same loop for the different APIs. This will also allow you to use C++ Lambda for unlinking.

navyxliu replied:
hi, Lam
Are you suggesting to use lambda+iterate() to implement unlink_all()? If we use `bucket` and `node` captured by the lambda, I think it's possible. However,  I think the resultant code is much less reasonable. Is it worth it? 

thanks,
--lx

iklam replied:
I mean we should implement a basic version of unlink() like this

```
template<typename Function>
void unlink(Function function) const { ...  }
```

Then `unlink_all` and  `template<class ITER> void unlink(ITER* iter)` can call into the above template.

navyxliu replied:
hi, @iklam @coleenp, 
I added `unlink(Function f)` per your request. Could you take a look?

coleenp commented on src/hotspot/share/utilities/resourceHash.hpp:
```diff
@@ -257,24 +279,32 @@ class ResourceHashtableBase : public STORAGE {
-        } else {
-          ptr = &(node->_next);
-        }
+    auto wrapper = [&](Node* const node, Node**& ptr) {
```
I wanted unlink to be lambda enabled, not an unlink_impl.  Can you just make unlink_all be something that provides a "return true" lambda and call unlink with that?

coleenp replied:
Or rather, first have a patch to add  unlink that takes a lambda and convert the uses to that.  I don't see the motivation for this change.

navyxliu replied:
hi, @coleenp 

My primary goal is 'unlink_all()'. I can create a Simple Iterator and use unlink() to implement that, but it would be sub-optimal(eg. [jvmtiTagMapTable::clear](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/jvmtiTagMapTable.cpp#L71)). When I roll it, I just realize that there are common logic among unlink(), unlink_all() and dtor. 'unlink_impl' is a template. It's not an API here. 

We can provide a lambda version of 'unlink()' in a new JBS issue, but I don't think we need to replace it with unlink(Iter*) completely. Functor and Lambda have different advantages. Functor is more formal and can be stateful. eg. 
`PurgeUnloadedConstraints` is big and deserve indents and comments. 'ResolutionErrorDeleteIterate' captures 'ConstantPool* p' for do_entry(). Can we do this is a separate JBS?

iklam replied:
My request is to make the unlink APIs use the same design pattern as the iterate APIs. You should be able to use either lambda or a function object for doing the unlink. There should be no performance penalty either way because the C++ compiler can optimize all the noise out.

iklam commented:
I've looked at the last version ([086750e](https://github.com/openjdk/jdk/pull/12213/commits/086750e46a4abb7115c2ff0ea49801c17df3981e)) in detail and I can understand why your implementation for `unlink` is so much more complex than `iterate`. The fundamental reason is sometimes we allow special clean up of the value (the old `unlink` API), while other times we don't (the destructor, and the new `iterate_all` API).

I think we need to rewrite ResourceHashtable to allow automatic and safe disposal of both the keys and the values. I've filed [JDK-8301296](https://bugs.openjdk.org/browse/JDK-8301296).

I would suggest putting this PR on hold until we come up with a new design that addresses the above issue.

iklam commented:
I updated [JDK-8301296](https://bugs.openjdk.org/browse/JDK-8301296) to clarify the problems with the current ResourceHashtable design and my proposal for fixing them.

I think the above proposed fixes shouldn't block the progress of this PR, which is just an optimization that maintains the current behavior. So let's continue the discussions here.

There are two parts of this PR:

[1] For the optimization (counting the number of entries and exit the loop early), do you have any performance data that shows this is beneficial?

For this optimization to be beneficial, we must have two conditions -- (a) the table is too large so it's likely to have many unused entries, and (b) the hash function is bad so most of the unused entries are at the end of the table.

For (a), maybe it's better to change the table to ResizeableResourceHashtable?
For (b), maybe you can also print out the occupancy of the table in your traces like this one (in your earlier PR https://github.com/openjdk/jdk/pull/12016)

```
[12.824s][debug][hashtables] table_size = 109, break at 68
```

If we have many entries (e.g., 40) but they all pack to the front end of the table, that means we have a bad hash function.

[2] As I suggested earlier, we should consolidate the code to use a single unlink loop, so you can apply this counting optimization in a single place. 

I am not quite sure why you would need the following in your "wrapper" functions:

```
if (clean) {
  *ptr = node->_next;
} else {
  ptr = &(node->_next);
}
```

and

```
if (node->_next == nullptr) {
  // nullify the bucket when reach the end of linked list.
  *ptr = nullptr;
}
```

I wrote a version of the consolidated loop that's much simpler. It also aligns with the old code so the diff is more readable:

https://github.com/openjdk/jdk/compare/master...iklam:jdk:8301136-resource-hash-unlink-all-suggestion

Note that I have both `unlink_all()` and `unlink_all(Function function)`, that's because the current API allows the user function to do two things (a) check if the entry should be freed, (b) perform special clean up on the entry. So if you want to free all entries but need to perform special clean up, you'd call  `unlink_all(Function function)`.

navyxliu commented:
hi, @iklam 

`unlink(Function f)` is the general API. It removes a selection of nodes. We don't know who is going to be deleted. To be conservative. we have to meticulously maintain the open-chain. `unlink_all()` knows that all nodes are all gone, so we don't need to maintain linked-list at all. 

Here is an example that an open-chain consists of 3 nodes. It explains why unlink_all() is better than 3 unlink() in a row. We don't need to maintain intermediate states such as `*ptr = &B`. 

![unlink_all_3_nodes](https://user-images.githubusercontent.com/2386768/215590069-fff2584f-55d0-4225-996e-44a2f4f0022f.png)

For `unlink_all()`, we need to update *bucket when we are at the last node of the open chain. 
That's that last step in the figure. 

```
if (node->_next == nullptr) {
  // nullify the bucket when reach the end of linked list.
  *ptr = nullptr;
}
```

The destructor version is even simplfied. Because the object is dead anyway, we just cut out all overhead.

navyxliu commented:
Performance-wise, I will evaluate hotspot and see if I can find support. thanks.

navyxliu commented:
hi, @iklam, 

I found that the dominating way to remove elements in resoureHashTable is its destructor. 
I ran some Renaissance benchmarks and observed the similar thing. 

In particular, it's very common to call destructor for an empty resourceHashtable. I think we should apply to the shortcut optimization in JDK-8300184 in the destructor. 


```
➜  jdk git:(JDK-8301136) ./build/linux-x86_64-server-release/images/jdk/bin/java -Xlog:hashtables=debug --version
[0.052s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.052s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.052s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.052s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.063s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.063s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.063s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.063s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.063s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.063s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.063s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.063s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.063s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.063s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.072s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.072s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
openjdk 21-internal 2023-09-19
OpenJDK Runtime Environment (build 21-internal-adhoc.xxinliu.jdk)
OpenJDK 64-Bit Server VM (build 21-internal-adhoc.xxinliu.jdk, mixed mode, sharing)
[0.073s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.073s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.073s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.073s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.073s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.073s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.073s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.073s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
[0.073s][debug][hashtables] ResourceHashtableBase destructor: entries = 0
[0.073s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed = 0
```

navyxliu commented:
I uploaded the hashtable logs of Renaissance/finagle-chirper. The reason I looked in it is that it produces the most logs in runtime. It yields 14,422 lines of log for the tag 'hashtables'.  One interesting thing is that 9,166 or 63.5% of them is to remove an empty ResourceHashable.

```
$awk -F '[ ,]' '$15 == 0 {print $0}' finagle-chirper-hashtables.log  | wc -l
9166
```

"break at 0" below indicates that we effectively skip any iteration.

```
[2.151s][debug][hashtables] ResourceHashtableBase table_size = 1031, break at 0, removed 0 out of 0
```

For the remaining non-trivial entries, 5254 out of 5256 is to remove all elements. They are all from destructor. Current patch only invokes `unlink_all()` in JVMTI. We don't use jvmti in the test. 

```
awk -F '[ ,]' '$12 == $15 {print $0}' finagle-chirper-hashtables-nontrivial.log|wc -l
5254
```

Overall, 14,063 or 97.5% of logs breaks earlier and save some useless iterations. 
```
 awk -F '[ ,]' '$5 != $9 {print $0}' finagle-chirper-hashtables.log   | wc -l
14063
```

Conclusion:
 1. We need to optimize 'destructor' of resourceHashtable. In particular, we should apply  the optimization of JDK-8300184 to it.
 2. we should consider to specialize "unlink_all".

navyxliu commented:
@iklam 

Like you said, I mixed up 2 things. My real intention is to introduce a new api `unlink_all()` because I used it in my project. 

I understand that you want me to refactor unlink(Iterator) using lambda. In order to have an efficient unlink_all() and dtor, I have to factor out unlink_impl(). It's private and acts as the algorithmic template.

What should I do now?  Should I split this JBS issue into 2? One is for the optimization and the other one for "unlink_all()"? that would make thing easier to review. 

thanks,
--lx

iklam commented:
> @iklam
> 
> Like you said, I mixed up 2 things. My real intention is to introduce a new api `unlink_all()` because I used it in my project.
> 
> I understand that you want me to refactor unlink(Iterator) using lambda. In order to have an efficient unlink_all() and dtor, I have to factor out unlink_impl(). It's private and acts as the algorithmic template.
> 
> What should I do now? Should I split this JBS issue into 2? One is for the optimization and the other one for "unlink_all()"? that would make thing easier to review.
> 
> thanks, --lx

I think it’s best to split this PR into two. We can first address the unlink_all issue and do the refactoring in a subsequent PR. Thanks
